pkg/insights/insightsreport/insightsreport.go

```diff
@@ -0,0 +1,164 @@
+	controllerstatus.Simple
+
+	configurator          Configurator
+	client                *insightsclient.Client
+	LastReport            types.SmartProxyReport
+	archiveUploadReporter <-chan struct{}
+	insightsReport        chan struct{}
+}
+
+// Response represents the Smart Proxy report response structure
+type Response struct {
+	Report types.SmartProxyReport `json:"report"`
+	Status string                 `json:"status"`
+}
+
+// Configurator represents the interface to retrieve the configuration for the gatherer
+type Configurator interface {
+	Config() *config.Controller
+	ConfigChanged() (<-chan struct{}, func())
+}
+
+// InsightsReporter represents an object that can notify about archive uploading
+type InsightsReporter interface {
+	ArchiveUploaded() <-chan struct{}
+}
+
+var (
+	// ErrWaitingForVersion contains an error when the cluster version is not loaded yet
+	ErrWaitingForVersion = fmt.Errorf("waiting for the cluster version to be loaded")
+
+	// insightsStatus contains a metric with the latest report information
+	insightsStatus = metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Namespace: "health",
+		Subsystem: "statuses",
+		Name:      "insights",
+		Help:      "Information about the cluster health status as detected by Insights tooling.",
+	}, []string{"metric"})
+)
+
+// New initializes and returns a Gatherer
+func New(client *insightsclient.Client, configurator Configurator, reporter InsightsReporter) *Gatherer {
+	return &Gatherer{
+		Simple:                controllerstatus.Simple{Name: "insightsreport"},
+		configurator:          configurator,
+		client:                client,
+		archiveUploadReporter: reporter.ArchiveUploaded(),
+	}
+}
+
+// PullSmartProxy performs a request to the Smart Proxy and unmarshal the response
+func (r *Gatherer) PullSmartProxy() {
+	klog.Info("Pulling report from smart-proxy")
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Minute)
+	defer cancelFunc()
+
+	klog.V(4).Info("Retrieving report")
+	config := r.configurator.Config()
+	reportEndpoint := config.ReportEndpoint
+	reportBody, err := r.client.RecvReport(ctx, reportEndpoint)
+	if authorizer.IsAuthorizationError(err) {
+		r.Simple.UpdateStatus(controllerstatus.Summary{
+			Operation: controllerstatus.DownloadingReport,
+			Reason:    "NotAuthorized",
+			Message:   fmt.Sprintf("Auth rejected for downloading latest report: %v", err),
+		})
+		return
+	} else if err != nil {
+		klog.Errorf("Error retrieving the report: %s", err)
+		return
+	}
+
+	klog.V(4).Info("Report retrieved")
+	reportResponse := Response{}
+
+	if err = json.NewDecoder(*reportBody).Decode(&reportResponse); err != nil {
+		klog.Error("The report response cannot be parsed")
+		return
+	}
+
+	klog.V(4).Info("Smart Proxy report correctly parsed")
+
+	if r.LastReport.Meta.LastCheckedAt == reportResponse.Report.Meta.LastCheckedAt {
+		klog.V(2).Info("Retrieved report is equal to previus one. Retrying...")
+		return
+	}
+
+	defer close(r.insightsReport)
```

Comment:
why deferring here, rather than just doing it as the last statement?

Reply:
Got it. I just copy/pasted it from the top of the function. Then I realized that I only want to close it when the function reach one specific point after some checks, so I just moved it

Reply:
Obviously, I will fix this one on my next commit